app/gui2/src/stores/graph.ts

```diff
@@ -5,46 +5,89 @@ import { Vec2 } from '@/util/vec2'
 import { assertNever, assert } from '@/util/assert'
 import { useProjectStore } from './project'
 import * as Y from 'yjs'
-import { useObserveYjs, useObserveYjsDeep } from '@/util/crdt'
+import { useObserveYjs } from '@/util/crdt'
 import {
-  rangeEncloses,
   type ContentRange,
   type ExprId,
   type IdMap,
   type NodeMetadata,
-} from '../../shared/yjs-model'
+  rangeIntersects,
+} from 'shared/yjs-model'
 import type { Opt } from '@/util/opt'
+import { parseEnso } from '@/util/ffi'
 
 export const useGraphStore = defineStore('graph', () => {
   const proj = useProjectStore()
 
   proj.setProjectName('test')
   proj.setObservedFileName('Main.enso')
 
-  let text = computed(() => proj.module?.contents)
-  let metadata = computed(() => proj.module?.metadata)
+  const text = computed(() => proj.module?.contents)
+  const metadata = computed(() => proj.module?.metadata)
+
+  const textContent = ref('')
 
   const nodes = reactive(new Map<ExprId, Node>())
   const exprNodes = reactive(new Map<ExprId, ExprId>())
 
   useObserveYjs(text, (event) => {
-    readState()
+    const delta = event.changes.delta
+    if (delta.length === 0) return
+
+    const affectedRanges: ContentRange[] = []
+    function addAffectedRange(range: ContentRange) {
+      const lastRange = affectedRanges[affectedRanges.length - 1]
+      if (lastRange?.[1] === range[0]) {
```

Comment:
probably would be good to confirm that either:
- ranges will never overlap, or
- we don't want to merge ranges that overlap